nuxeo-services/nuxeo-permissions/src/main/java/org/nuxeo/ecm/permissions/PermissionGarbageCollectorListener.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * (C) Copyright 2023 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.nuxeo.ecm.permissions;
+
+import static org.nuxeo.ecm.core.api.event.DocumentEventTypes.DOCUMENT_REMOVED;
+import static org.nuxeo.ecm.permissions.Constants.ACE_INFO_DIRECTORY;
+
+import java.util.Collections;
+import java.util.Map;
+
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.event.Event;
+import org.nuxeo.ecm.core.event.EventContext;
+import org.nuxeo.ecm.core.event.EventListener;
+import org.nuxeo.ecm.core.event.impl.DocumentEventContext;
+import org.nuxeo.ecm.directory.Session;
+import org.nuxeo.ecm.directory.api.DirectoryService;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.services.config.ConfigurationService;
+
+/**
+ * Listener cleaning the 'aceinfo' directory when a document is removed.
+ *
+ * @since 1010-HF70
+ */
+public class PermissionGarbageCollectorListener implements EventListener {
+
+    protected static final String ACE_INFO_GC_ENABLED_PROPERTY = "nuxeo.aceinfo.gc.enabled";
+
+    @Override
+    public void handleEvent(Event event) {
+        EventContext ctx = event.getContext();
+        if (!(ctx instanceof DocumentEventContext)) {
+            return;
+        }
+
+        ConfigurationService configService = Framework.getService(ConfigurationService.class);
+        if (DOCUMENT_REMOVED.equals(event.getName())
```

Comment:
I guess there is no way to know if the removed document hold and ACP and require a cleaning?

Reply:
At this point if I want to check the doc `ACP`:
```((DocumentEventContext) ctx).getSourceDocument().getACP();```
I get a `DocumentNotFoundException`
```2023-01-06 12:14:26,335 [main] ERROR [EventServiceImpl] Exception during permissionGarbageCollectorListener sync listener execution (after 1ms), continuing to run other listeners
org.nuxeo.ecm.core.api.DocumentNotFoundException: 8d1a455f-f14f-4390-bbad-b7218ca83c76
	at org.nuxeo.ecm.core.storage.sql.coremodel.SQLSession.getDocumentByUUID(SQLSession.java:185) ~[classes/:?]
	at org.nuxeo.ecm.core.api.AbstractSession.resolveReference(AbstractSession.java:340) ~[classes/:?]
	at org.nuxeo.ecm.core.api.AbstractSession.getACP(AbstractSession.java:570) ~[classes/:?]
	at org.nuxeo.ecm.core.api.impl.DocumentModelImpl.getACP(DocumentModelImpl.java:759) ~[classes/:?]
	at org.nuxeo.ecm.permissions.PermissionGarbageCollectorListener.handleEvent(PermissionGarbageCollectorListener.java:53) ~[classes/:?]```

I don't know if there is another way to find out.

Reply:
yes this is expected, the listener is async so the doc is already removed, just wondering if there is another event or if it's worth creating an additional event like ACP_REMOVED